doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,206 @@
+=========
+Pipelines
+=========
+
+Background
+==========
+
+Requests
+--------
+
+OSD serves different kinds of requests. Some of them are relatively easy to handle. For instance,
+the "ceph config get" command is directed straight to the config subsystem, and is fulfilled
+right away. But some are more "demanding". The most popular two of them are probably:
+
+- I/O requests from clients.
+- PG peering requests from peer OSDs.
+
+Because inherently a distributed storage system is distributed by its own nature, we can not assume
+that an OSD or the PGs hosted by it are always ready to serve a request targeting a shared state, not
+to mention the unpredictable hardware failures. But fortunately, even if an OSD is not able to serve
+a request right away, under most circumstances, it can postpone the request, prepare itself so that it
+can answer this request. It's quite common in daily life:
+
+> Greate question, let me look into it and I will get back to you!
+
+Some questions involves multiple steps. To serve an I/O request from client, at
+least 7 stages are involved:
+
+.. ditaa::
+
+   +----+---------------------+
+   |cGRE| OSD wait for osdmap |
+   +----+----+----------------+
+             |
+             V
+   +----+-----------------+
+   |cGRE| OSD wait for PG |
+   +----+-----------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for osdmap |
+   +----+--------------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for active |
+   +----+--------------------+
+             |
+             V
+   +----+----------------------+
+   |cBLU| PG wait for recovery |
+   +----+----------------------+
+             |
+             V
+   +----+---------------------+
+   |cBLU| wait for object ctx |
+   +----+---------------------+
+             |
+             V
+   +----+------------------+
+   |cBLU| wait for process |
+   +----+------------------+
+
+             legend
+             +---------------+
+             |cGRE OSD stage |
+             +---------------+
+             |cBLU PG stage  |
+             +---------------+
+
+Two Restrictions
+----------------
+
+An OSD client talks with each connected OSD over an TCP connection. The client is allowed to
+perform asynchronous I/O with OSD in the sense that it can send another request without waiting
+for the completion of the previous one. But the contract of the RADOS protocol requires the OSD
+to reply the requests in the exact order in which those requests are received from the same
+client. In other words, OSD is not allowed to reoder the requests even if they are all read
```

Comment:
...
requres writes (and write ordered reads) performed on a single object to be performed in the order in which they were originally submitted, and only once.
...

Reply:
Note, this is *not* a property of the OSD, it's a property of librados as a whole and is implemented by code located on the client, within the client and server portions of the messenger, and on the osd.

Reply:
And it actually can reorder read requests even on the same object, there's just little point to doing so.

Reply:
If we thought that we'd actually get enough reads on the same pg in parallel to justify the complexity, we actually could permit overlapping reads within a pg or a single object.